[CodeGen] Use first EHLabel as a stop gate for live range shrinking

MuellerMP opened the pull request:
This fixes issue #114194

The issue happens during the `LiveRangeShrink` pass, which runs early, before phi elimination. LandingPads, which are lowered to EHLabels, need to be the first non phi instruction in an EHPad. In case of a phi node being in front of the EHLabel and a use being after the EHLabel, we hoist the use in front of the label.

This results in a portion of the landingpad missing due to being hoisted in front of the label.

MuellerMP committed:
[CodeGen] Use first EHLabel as a stop gate for live range shrinking

This fixes issue #114194

The issue happens during the LiveRangeShrink pass, which runs early, before phi elimination. LandingPads, which are lowered to EHLabels, need to be the first non phi instruction in an EHPad. In case of a phi node being in front of the EHLabel and a use being after the EHLabel, we hoist the use in front of the label.

This results in a portion of the landingpad missing due to being hoisted in front of the label.

MuellerMP commented:
An alternative solution could be eliminating phis before lrshrink, which would lower the phis into copies after the label.

arsenm reviewed (commented):
What is the context for this pass? I didn't know it existed, and I see it's only added by x86. Why does it exist? This is just an extra machine scheduler?

arsenm commented on llvm/lib/CodeGen/LiveRangeShrink.cpp:
```diff
@@ -125,14 +135,19 @@ bool LiveRangeShrink::runOnMachineFunction(MachineFunction &MF) {
       MachineInstr &MI = *Next;
       ++Next;
+
       if (MI.isPHI() || MI.isDebugOrPseudoInstr())
         continue;
+      if (IsEHPad && !EHBarrier && MI.isEHLabel())
```
Should this be using SkipPHIsLabelsAndDebug?

MuellerMP replied:
Well it could have been using it if we did not want to check for labels but in my fix i use the first ehlabel to determine the start of the landingpad.

MuellerMP commented:
> What is the context for this pass? I didn't know it existed, and I see it's only added by x86. Why does it exist? This is just an extra machine scheduler?

That's a good question, because the machine scheduler actually doesn't run into this problem due to running after phi elimination.
So I read the original [MR](https://reviews.llvm.org/D32563) and this pass seems to relieve register pressure in some special cases: 
1) when reassociation reorders ir instructions in a bad way
2) when user input introduces live-range issues
It also seemd to require insight of the whole basicblock, which is why it couldn't be solved with exisiting generic schdeduling.

I wonder if these assumptions still hold in the current version of misched and resassociate.
Since I'm not to familiar with either of them I propose to fix the current version of the pass for now and leave the analysis regarding the redundancy of the pass for followups.

MuellerMP commented:
Updated the patch to deal with rebuilds of the InstructionOrderMap (triggered an assertion during my tests).
I now use the same method the original author used for code motion barriers related to instruction with side effects:
Rebuilding the InstructionOrderMap so illegal code motion insertion points can't be referenced with FindDominatedInstruction.

MuellerMP commented on llvm/lib/CodeGen/LiveRangeShrink.cpp:
```diff
@@ -135,6 +137,13 @@ bool LiveRangeShrink::runOnMachineFunction(MachineFunction &MF) {
+        // barrier for code motion. IOM is rebuild from the next instruction
+        // to prevent later instructions from being moved before this MI.
+        SawEHLabel = true;
+        BuildInstOrderMap(Next, IOM);
```
Ah yes - I think I get what you were hinting at now =)
So i cant SkipPHIsLabelsAndDebug in every case since we will miss the optimization in the [original test](https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/lrshrink.ll) which relies on moving uses of phis to their defs.
But we should be able to utilize the skip in the EHPad case since the assumption of ehlabels being the first non phi should still hold.

arsenm reviewed (approved):
Not really sure why you can't do this in the !MBBPad case

MuellerMP commented:
> Not really sure why you can't do this in the !MBBPad case

well not for the map at least -> otherwise we are missing the phis in the map and thus we can't hoist their uses

MuellerMP commented:
Updated the patch again:
- Added a comment explaining the initial SkipPHIsLabelsAndDebug call for EHPads
- Added a iterator-equals-end check before the initial BuildInstOrderMap for EHPads due to BuildInstOrderMap derefing the iterator
- Now using SkipPHIsLabelsAndDebug infront and inside the loop thus eliminating the if continue

Did further testing with this patch and it seems to achieve the desired result.
I hope this clarifies my fix and also incorporates the adjustments regarding the Skip func in a meaningful way.

@arsenm would you be ok with merging this for me since i don't have write access?

nathanchance commented:
I am noticing a significant compile time regression (perhaps infinite loop?) with this change with certain compile options, which I discovered when building the Linux kernel for 32-bit x86. `cvise` spits out the trivial C reproducer:

```
enum { RATE_6M_PLCP } il4965_tx_status_reply_tx_agg_2;
int il4965_tx_status_reply_tx_i;
short il4965_tx_status_reply_tx_start_idx,
    il4965_tx_status_reply_tx_frame_status_0_0;
int il4965_tx_status_reply_tx() {
  int sh, start = il4965_tx_status_reply_tx_start_idx;
  long long bitmap;
  for (; il4965_tx_status_reply_tx_i; il4965_tx_status_reply_tx_i++) {
    if (il4965_tx_status_reply_tx_frame_status_0_0)
      continue;
    if (sh > 4) {
      sh = start;
      bitmap = sh = 0;
    }
    bitmap |= 1ULL << sh;
  }
  if (bitmap)
    il4965_tx_status_reply_tx_agg_2 = 1;
  return 0;
}
```

Immediately prior to this change:

```
Benchmark 1: clang --target=i686-linux-gnu -march=geode -O2 -g -c 4965-mac.i
  Time (mean ± σ):      14.9 ms ±   0.3 ms    [User: 8.9 ms, System: 5.9 ms]
  Range (min … max):    14.5 ms …  17.0 ms    142 runs
```

At this change:

```
$ clang --target=i686-linux-gnu -march=geode -O2 -g -c 4965-mac.i
<hangs>
```

MuellerMP commented:
@nathanchance ty for finding and reducing this issue! 
can repro and have a fix ready.
